orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -921,35 +921,39 @@ IMPLEMENT_GRADIENT_BUILDER(GetFastGeluGradient) {
 }
 
 IMPLEMENT_GRADIENT_BUILDER(GetSendGradient) {
-  NodeAttributes new_attrs;
-  const auto& old_attrs = SrcNodeAttributes();
- 
-  new_attrs["src"] = MakeAttribute("src", old_attrs.at("dst").i());
-  new_attrs["dst"] = MakeAttribute("dst", old_attrs.at("src").i());
-  new_attrs["tag"] = MakeAttribute("tag", old_attrs.at("tag").i());  // keep tag the same for now
-  new_attrs["element_type"] = MakeAttribute("element_type", old_attrs.at("element_type").i());
+  // Send inputs: signal A, remote, data; outputs: signal B
+  // Recv inputs: signal B, remote; outputs: signal A', data'
+
+  std::vector<ArgDef> out_args;
+  out_args.push_back(GI(0));  // Signal
+  for (int i = 2; i < GetSrcNodeInputSize(); ++i) {
+    out_args.push_back(GI(i));  // Data
+  }
 
   return std::vector<NodeDef>{
       NodeDef("Recv",
-              {O(0)},
-              {GI(0), GI(1)},
-              new_attrs)};
+              {O(0), I(1)},  // {Signal, Remote}
+              out_args,
+              SrcNodeAttributes())};
 }
 
 IMPLEMENT_GRADIENT_BUILDER(GetRecvGradient) {
-  NodeAttributes new_attrs;
-  const auto& old_attrs = SrcNodeAttributes();
+  // Recv inputs: signal A, remote; outputs: signal B, data
+  // Send inputs: signal B, remote, data'; outputs: signal A'
 
-  new_attrs["src"] = MakeAttribute("src", old_attrs.at("dst").i());
-  new_attrs["dst"] = MakeAttribute("dst", old_attrs.at("src").i());
-  new_attrs["tag"] = MakeAttribute("tag", old_attrs.at("tag").i());  // keep tag the same for now
-  new_attrs["element_type"] = MakeAttribute("element_type", old_attrs.at("element_type").i());
+  std::vector<ArgDef> in_args;
+  in_args.push_back(O(0));  // Signal
+  in_args.push_back(I(0));  // Remote
```

Comment:
If Recv has src=10, would this mean Send has dst=10? From my impression, the src and dst are likely two different numbers in MPI.

Reply:
They are different number, and one of them should be its own rank.
In backward, it holds.

Reply:
I can see your point now.